lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -302,26 +372,67 @@ void SessionStateEnforcer::m5g_complete_termination(
   return;
 }
 
-/*Removing all associated rules to session*/
-void SessionStateEnforcer::m5g_remove_rules_and_update_upf(
-    const std::string& imsi, const std::unique_ptr<SessionState>& session,
-    const std::string& dnn, SessionStateUpdateCriteria& uc) {
-  // remove all rules;
-  MLOG(MINFO) << "Will be removing all associated rules of session id "
+void SessionStateEnforcer::m5g_state_change_action(
+    std::string& imsi, SessionStateUpdateCriteria& session_uc,
+    SetSmNotificationContext notif, std::unique_ptr<SessionState>& session) {
+  /*Irrespective of any State of Session, release and terminate*/
+  if ((notif.common_context().sm_session_state() == ACTIVE_2) &&
+      (session->get_state() == INACTIVE)) {
+    m5g_execute_state_change_action(session_uc, session, ACTIVE);
+    /* Call for all rules to be associated from session
+     * and inform to UPF
+     */
+    m5g_pdr_rules_change_and_update_upf(imsi, session, PdrState::INSTALL);
+  }
+  /* Now code for Inactive state notificaiton
+   */
+  else if (
+      (notif.common_context().sm_session_state() == INACTIVE_3) &&
+      (session->get_state() == ACTIVE)) {
+    m5g_execute_state_change_action(session_uc, session, INACTIVE);
+    /* Call for all rules to be de-associated from session
+     * and inform to UPF
+     */
+    m5g_pdr_rules_change_and_update_upf(imsi, session, PdrState::IDLE);
+  }
+}
+
+void SessionStateEnforcer::m5g_execute_state_change_action(
+    SessionStateUpdateCriteria& session_uc,
+    std::unique_ptr<SessionState>& session, SessionFsmState targetState) {
+  auto stateStr = session_fsm_state_to_str(targetState);
+  MLOG(MINFO) << stateStr << " state transition for session id "
               << session->get_session_id();
-  auto ip_addr = session->get_config()
-                     .rat_specific_context.m5gsm_session_context()
-                     .pdu_address()
-                     .redirect_server_address();
+  session->set_fsm_state(targetState, session_uc);

```

Comment:
The `set_fsm_state` already logs state transitions so wouldn't these logs be noisy?

Reply:
This function also seems like it could be owned by `SessionState`.

Reply:
we will remove it before final commit for sure. Some more idle mode testing pening.